### PR TITLE
Add integration tests to produce and fetch messages

### PR DESCRIPTION
**Pull request by Hackzzila**
There are a few things in this PR, I can split it up if you would like.

- Derive `PartialEq` and `Eq` for `Record` (and `TimestampType`)
- Move from `docker-compose-runner` to `testcontainers` to make it easier to spawn one container per test. I couldn't find a good way to do this with `docker-compose-runner`, but I am open to suggestions
- Renames `integration_test` to `get_api_versions`
- Adds 2 new tests:
  - `record_batch_produce_fetch` encodes 3 records as a v2 record batch, sends them to kafka, and fetches them back
  - `message_set_v1_produce_fetch` does the same but for a v1 message set. This currently fails, and is fixed in #62

### CODE REVIEW

**Comment by Hackzzila** [comment]
CI failed because of `message_set_v1_produce_fetch` which is expected

**Comment by rukai** [comment]
Thankyou so much for this.
Happy to swap out the Docker library, I'll look into what test containers brings since I've not used it before.

I'll do a proper review on Monday, in the meantime might be a good idea to assert the expected bad behaviour with a TODO so that we can land this PR. And then we'll change it to assert the correct behaviour when we land the fix.

File: tests/integration_tests.rs [hunk 1]
```diff
@@ -1,21 +1,15 @@
...
-#[test]
-fn integration_test() {
-    let _docker_compose = docker_compose("tests/docker-compose.yaml");
+mod common;
+use common::*;
 
-    let mut socket = TcpStream::connect("localhost:9092").unwrap();
+#[test]
+fn get_api_versions() {
+    let container = start_kafka();
```
Comment (tychedelia) [inline tests/integration_tests.rs:11]: nit: we should consider just wrapping these up in a `setup` fn instead.

File: tests/produce_fetch.rs [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            compression: Compression::None,
+        },
+    )
+    .unwrap();
+
+    create_topic(topic_name.clone(), &mut socket);
+
+    // Sometimes (rarely) the produce request will fail with error code 6 (NOT_LEADER_OR_FOLLOWER)
+    // Hopefully a 1s sleep will prevent that
+    std::thread::sleep(Duration::from_secs(1));
```
Comment (tychedelia) [inline tests/produce_fetch.rs:50]: if this ends up being flaky, we could probably do a topic list in a loop to wait for the node to elect itself?

**Review by tychedelia — APPROVED** [review]
Thanks, very helpful to expand our test footprint. Will defer to @rukai's approval on the changes to the container runner, to which I'm agnostic as long as it runs green in CI.

**Comment by tychedelia** [comment]
All green, thanks @Hackzzila!

File: tests/produce_fetch.rs [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        Compression, Record, RecordBatchDecoder, RecordBatchEncoder, RecordEncodeOptions,
+        TimestampType,
+    },
+};
+
+mod common;
+use common::*;
+
+#[test]
+fn record_batch_produce_fetch() {
```
Comment (rukai) [inline tests/produce_fetch.rs:23]: Considering these two tests are variants of each other, maybe name this one as `message_set_v2_produce_fetch`?

**Review by rukai — APPROVED** [review]
These tests are so good, thanks!

**Pull request merged**